FixNullable is breaking ability to validate against api doc

We're using a gateway API and various clients will connect to it to fetch the json api and will then generate clients sdks.  We've noticed that the first time it works, but after the API is used, it no longer works, failing with the following error from oas-validator:

```
{ AssertionError: expected Array [ 'boolean', 'null' ] to have type string
```

Comparing the two docs (one that works, the other that doesn't) it appears the `type` of a few properties has been mutated from a string (`'boolean'`) to an array (`['boolean', 'null']`)

This bug was introduced here:  e699a14

According to `oas-validator`, the `type` property cannot be an array, only a string:
https://github.com/Mermade/oas-kit/blob/master/packages/oas-validator/schemas/openapi-3.0.json#L410  

Further, "null" is not a valid value here.  It does specify `nullable` in that same schema.  It seems the proper fix for nullable is to check against the `nullable` property in the spec (if it exists) and to do the check based on that.

I'm not sure how well-suited `z-schema` is for this.  Serendipitously it seems to take a schema that is _very_ close to how the openapi schema looks, with this difference it seems -- only a single type for a property is allowed, and nullable is defined by the `nullable` property.

It could be this `fixNullable` could go inside `oas-validator.js` instead -- perform the same change to the schema to make it `z-schema` compatible.  One would still need to take care to create a new schema for z-schema with the changes to type, and to *not* mutate the schema that is swagger-specific.

--- a/src/lib/utils.js
+++ b/src/lib/utils.js
@@ -22,6 +22,16 @@ Considerar que:
 
 var validator = require('validator');
 
+var fixNullable = function(schema) {
+  Object.getOwnPropertyNames(schema).forEach((property) => {
+    if (typeof schema[property] === 'object') {
+      fixNullable(schema[property]);
+    } else if (property === 'type' && typeof schema[property] === 'string' && schema.nullable === true) {
+      schema.type = [schema.type, "null"];
+    }
+  });
+}
+
 /**
  * Generates a valid name, according to value of nameFor.
  * @param {string} input - String to generate a name from.
@@ -47,5 +57,6 @@ var generateName = function(input, nameFor) {
 }
 
 module.exports = {
-  generateName: generateName
+  generateName: generateName,
+  fixNullable: fixNullable
 };

--- a/src/middleware/oas-router.js
+++ b/src/middleware/oas-router.js
@@ -21,6 +21,7 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.*/
 
 var exports; // eslint-disable-line
 var path = require('path');
+var _ = require('lodash-compat')
 var ZSchema = require("z-schema");
 var MIMEtype = require('whatwg-mimetype');
 var config = require('../configurations'),
@@ -32,16 +33,6 @@ var validator = new ZSchema({
 });
 var utils = require("../lib/utils.js");
 var controllers; // eslint-disable-line
-
-function fixNullable(schema) {
-  Object.getOwnPropertyNames(schema).forEach((property) => {
-    if (typeof schema[property] === 'object') {
-      fixNullable(schema[property]);
-    } else if (property === 'type' && typeof schema[property] === 'string' && schema.nullable === true) {
-      schema.type = [schema.type, "null"];
-    }
-  });
-}
 
 function getExpectedResponse(responses, code) {
   // Exact match wins over range definitions (1XX, 2XX, 3XX, 4XX, 5XX)
@@ -150,8 +141,9 @@ function checkResponse(req, res, oldSend, oasDoc, method, requestedSpecPath, con
     }
     if (resultType && resultType.essence === 'application/json') {
       //if there is no content property for the given response then there is nothing to validate.
-      var validSchema = responseCodeSection.content['application/json'].schema;
-      fixNullable(validSchema);
+      var validSchema = _.cloneDeep(responseCodeSection.content['application/json'].schema);
+      utils.fixNullable(validSchema)
+
       content[0] = JSON.stringify(content[0]);
       logger.debug("Schema to use for validation: " + JSON.stringify(validSchema));
       var err = validator.validate(data, validSchema);

--- a/src/middleware/oas-validator.js
+++ b/src/middleware/oas-validator.js
@@ -30,7 +30,8 @@ var ZSchema = require("z-schema");
 // var urlModule = require('url');
 
 var config = require('../configurations'),
-  logger = config.logger;
+  logger = config.logger,
+  utils = require('../lib/utils');
 var validator = new ZSchema({
   ignoreUnresolvableReferences: true,
   ignoreUnknownFormats: config.ignoreUnknownFormats,
@@ -115,7 +116,9 @@ function checkRequestData(oasDoc, requestedSpecPath, method, res, req, next) { /
       } else {
         // can be any of "application/json", "multipart/form-data", "image/png", ...
         const contentType = Object.keys(requestBody.content)[0];
-        var validSchema = requestBody.content[contentType].schema;
+        var validSchema = _.cloneDeep(requestBody.content[contentType].schema)
+        utils.fixNullable(validSchema)
+
         var data = req.body; //JSON.parse(req.body); //Without this everything is string so type validation wouldn't happen TODO: why is it commented?
         // a multipart/form-data request has a "files" property in the request whose
         // properties need to be passed to evaluating the required parameters in the openAPI spec

